usaspending_api/etl/elasticsearch_loader_helpers/aggregate_key_functions.py

```diff
@@ -0,0 +1,181 @@
+import json
+import logging
+
+from typing import Optional, List
+
+
+logger = logging.getLogger("script")
+
+
+def award_recipient_agg_key(record: dict) -> str:
```

Comment:
For all agg key functions here -- 
We previously talked about optimization techniques, one of which you mentioned was memoization ... which I believe I read as caching in memory of the python interpreter the output of a function given inputs. 

Since we pass all ES docs through here, and the idea of an agg key is there are likely to be many docs with the same key ... it may have good effect here (if it doesn't bloat memory with so many entries). 

I also came across Python's `@functools.lru_cache(maxsize=None)` decorator which _i think_ is basically short-hand for doing memoization. 

https://docs.python.org/3.7/library/functools.html

Some caveats that may need further testing/experimentation/extension:
1. With maxsize=None we'd have to watch for memory bloating
1. LRU = least-recently-used, which means it will keep the recent func call args, assuming it's a better predictor of what's to-come. That's not necessarily the case. Maybe we could come up with an alternative eviction that keeps the entries with the highest count?

Reply:
Looks like `cachetools` lib is a good start for exploration of specific strategies (if memory becomes problematic): https://cachetools.readthedocs.io/en/stable/

Reply:
I'll have to look into cachetools. The basic lru_cache does not allow caching a mutable function parameter like a dictionary.